### DISCUSSION TITLE
Call for testing the new coming soon mode

### DISCUSSION THREAD

**Discussion by adrianduffell** [woocommerce/woocommerce#47418] [Coming Soon Mode]
We are planning to introduce a new coming soon mode to WooCommerce in an upcoming version. This aims to allow new merchants to build their store in private, while growing their online presence with a coming soon page. It is coupled with a new "Launch Your Store" setup task which aims to give users more confidence taking their store live.

<p align=center><img width="500" alt="coming soon landing page" src="https://github.com/woocommerce/woocommerce/assets/9312929/ab4df6b8-a052-4a9c-8115-5d56f420c21e"></p>

## What to test

If you have a moment, we’d love to hear your feedback! Specifically, we’d love to hear how you find using the new "Site visibility" settings tab and templates for the coming soon mode. The [nightly build of WooCommerce](https://github.com/woocommerce/woocommerce/releases/tag/nightly) contains the feature. Just upload `woocommerce-trunk-nightly.zip` into a suitable test environment or spin up a [WP Playground](https://bit.ly/woo-coming-soon-mode)!

Once the nightly is uploaded, a new badge appears on the WooCommerce Admin home screen, indicating the status of the site and providing a dropdown to quickly access these features.

<p align=center><img width="500" alt="discussion -badge" src="https://github.com/woocommerce/woocommerce/assets/9312929/f28863c0-011e-46a9-af66-fccf13e5bf34"></p>

## Site visibility settings

The coming soon mode is controlled by a new tab in WooCommerce settings. 

<p align=center><img width="500" alt="discussion - site visibility settings" src="https://github.com/woocommerce/woocommerce/assets/9312929/f560f477-9a9b-4c3b-b48b-c7cf98f32d6c"></p>

When in coming soon mode, additional toggles appear. 

* **Restrict to store pages only**. This is useful for merchants installing WooCommerce on an existing site. Only the WooCommerce store-related pages will be protected by the coming soon mode.
* **Share your site with a private link**. This allows merchants to share the progress of their new store with friends, family, and colleagues without them needing to login as an admin user.

Existing stores will default to live mode, while new stores that complete the profiler will default to coming soon mode. The following logic dictates whether the coming soon protection applies to the whole site or is restricted to store pages only:

* Default to whole site, If the site is fresh (as determined by the [fresh_site WordPress flag](https://core.trac.wordpress.org/ticket/38533)), and the first admin user registration date is less than a month ago.
* Otherwise, enable the "Restrict to store pages only" toggle.

Please try switching between live and coming soon modes. Does the front-end change appropriately? If not, what hosting setup do you have? Note: you will need to view the front-end as a non-admin user (use a private window or another browser).

## Coming soon templates

Don't like the purple? The coming soon page template is fully-customizable in the WordPress Editor. 

<p align=center>
<img width="500" alt="coming soon template" src="https://github.com/woocommerce/woocommerce/assets/9312929/c4fcb746-0925-4d63-8d29-6631b0a29a81">
</p>

How easy/difficult do you find editing this template? We intend for the editing experience to be as close to 1:1 with the front-end as possible. If you find something that doesn't render properly on the front-end, let us know and state the theme being used.

Additionally, when the "Restrict to store pages only" toggle is enabled, the template will adjust to the active theme and render a header or footer.

<p align="center"><img width="500" alt="discussion - template restricted" src="https://github.com/woocommerce/woocommerce/assets/9312929/d17eb23a-4ca0-4727-85bf-e42697101bdb"></p>

## Known issues

* There may be conflicts with other coming soon modes provided by hosts and other plugins.

## Over to you!

Thanks for contributing and helping to shape this feature! All feedback is much appreciated.

**Comment by Brianmitchtay** [comment]
Everything worked as expected for me in my test on an established localhost install using WPLocal to run the site. 

Having any Woo page (e.g. cart, checkout) go to the coming soon template is pretty slick. Excited to see this ship!

As far as editing the template goes it took me a while to figure out that the background color was not set as an actual style, but was a **setting** on the "Coming Soon" block. I was expecting to find the setting in the styles for the coming soon block, not the settings for it. But I found it eventually all the same.

![Screenshot 2024-05-13 at 4 36 11 PM](https://github.com/woocommerce/woocommerce/assets/45057772/3dfb1cbb-2e2f-42fa-8c41-f6acf485dc0d)

**Comment by webdados** [comment]
I don't think "Site visibility" needs its own tab on WooCommerce settings. Tabs are premium real estate.
Two options could be easily integrated into the General tab or as a sub-section of Advanced.

**Reply by jamesckemp** [reply]
We decided to give it its own tab because of a few reasons (we discussed it extensively and created a few design scenarios 😅):

1. The settings are using a dynamic implementation and newer design. When added to the "General" tab, they sat out of line with all the other settings and looked odd. We considered adding a "Site Visibility" label to push them inline, but it just meant duplicating labels and added confusion.
2. When using the WooCommerce menu item to "Manage Visibility", there was no reliable way to locate the correct settings when they were mixed among the other "General" settings.

I was also reluctant to add an additional tab, but it gave the best experience in the end. However, happy to hear any ideas you might have to combat the two issues mentioned above?

**Reply by webdados** [reply]
I don't have ideas to tackle the two issues mentioned, and it seems you already did that work. It just seems a waste of premium space to have a tab just for this.

**Reply by jamesckemp** [reply]
I would expect the whole settings area to be revamped at some point, more in line with this new setting. I'll definitely keep a look out for similar feedback to see if we need to revisit the decision, though.

**Comment by cre8tivediva** [comment]
Is "Site" the entire site or the "Store". If the Store, I'd recommend changing to "Store Visibility". I also agree with @webdados to pull it off the Settings tab. I wouldn't want them to "hunt" for it. But with most "maintenance" type plugins that I've worked with, they have it as a toggle on the Admin Bar across the top. And, at the two additional arrows, they reference "site" vs. "store". If it's really the store, having "site" could be confusing.

<img width="1792" alt="330088994-f560f477-9a9b-4c3b-b48b-c7cf98f32d6c" src="https://github.com/woocommerce/woocommerce/assets/4140035/88628e3f-f7a5-4b73-b990-3607e5265e64">

**Reply by jamesckemp** [reply]
It's the entire site unless you enable "Restrict to store pages only", then it's only the store!

**Reply by cre8tivediva** [reply]
Hummm, that's great to know then so I can test with other maintenance plugins that I have installed.

**Reply by adrianduffell** [reply]
Thanks @cre8tivediva , I would love to know the results of your testing with other maintenance plugins once you're done. We plan to avoid clashes with these next.

**Reply by cre8tivediva** [reply]
@adrianduffell sure! I have two more I can test on this one.

**Comment by alexstandiford** [comment]
It would also be nice to set this up to optionally redirect to an existing page instead. That way if your home page is the same landing page as what you'd want people to see for this "coming soon" page, it would work.

PS: I would be using this _right now_ if I could. Such a great idea.

**Reply by cre8tivediva** [reply]
I also agree with @alexstandiford . What would the response code be on coming soon if it's not redirected? How would that be handled between "site" and "store"?

**Reply by jamesckemp** [reply]
That's a good question for @adrianduffell to answer!

**Reply by psealock** [reply]
Thats a great idea @alexstandiford. @adrianduffell, I think we can add a filter here so folks can redirect to a template of their choice. I'm not sure off the top of my head how you might redirect to an existing page, but surely its possible.

https://github.com/woocommerce/woocommerce/blob/36fa4f82edd1f097c3ee2f929486c57285d78314/plugins/woocommerce/src/Internal/ComingSoon/ComingSoonRequestHandler.php#L44-L46

@cre8tivediva the response code on the coming soon pages are 200 after they are redirected. It won't be possible to access site or store pages in coming soon mode. Does that answer your question?

**Reply by cre8tivediva** [reply]
@psealock yes, that's great on response code.

**Reply by alexstandiford** [reply]
> Thats a great idea @alexstandiford. @adrianduffell, I think we can add a filter here so folks can redirect to a template of their choice. I'm not sure off the top of my head how you might redirect to an existing page, but surely its possible.

I would probably do it using [get_post_permalink](https://developer.wordpress.org/reference/functions/get_post_permalink/) and [wp_redirect](https://developer.wordpress.org/reference/functions/wp_redirect/).

Does the request handler run inside `template_redirect`? If so, I think what I'm suggesting would work fine.

**Comment by cre8tivediva** [comment]
Another suggestion that might not be as exhausting to create the Coming Soon page, would be to allow users to create their own and select it like they do the remainder of the pages. Maybe put a "Coming Soon" or "Site Visibility" area on that page. This way if they want to create customized pages at various times or swap them out, they'd only need to create it and then set the page here. But there would be a default like what's there now but add this so they have an option.

![Screenshot 2024-05-22 at 15-29-40 WooCommerce settings ‹ Savannah Demo — WordPress](https://github.com/woocommerce/woocommerce/assets/4140035/b1d86b8c-3f91-4dac-9c12-b1a0fb0b3020)

**Comment by cre8tivediva** [comment]
I did a walk through via video - https://www.screencast.com/t/YykSpLfi

On **Site Visibility** - Coming Soon, I'd would have that open because as you can see from my video, I was confused. What I was seeing didn't look like the image above. There are quite a few users who might also get confused if they do not see the toggles at first glance when they go to the Settings page.

![2024-05-22_17-11-12](https://github.com/woocommerce/woocommerce/assets/4140035/d8379c32-a388-46be-ace1-c95cc620cf7a)


<hr>
How can the message on the Shop page be modified? I know from my own audience of users, that will be the first thing they look for or send a message seeking how to change. I offer tech support and provide theme support and I can pretty much bet that will be one of the first questions a user will contact support for.

![2024-05-22_17-12-16](https://github.com/woocommerce/woocommerce/assets/4140035/c1973fe7-318f-4f92-b2ea-6cfe46efefa9)


<hr>
When setting the "Shop only" to Coming Soon mode, if an existing site has products on the front or WooCommerce products blocks scattered around, i.e., on other pages, sidebar, etc., they'll need to be reminded to remove that section from their site. I was expecting that area to disappear or render some sort of message on the block, not stay Live. At the 3:28 mark I said this section was Kadence but it's not. It's the default "Newest Products" Block in WooCommerce. So if that was left there and they turned on Coming Soon mode, shouldn't these products disappear? Shouldn't there be a message like on the Shop page?

![2024-05-22_17-27-39](https://github.com/woocommerce/woocommerce/assets/4140035/bc5cee45-10c2-498e-9919-92c539b33a32)

**Reply by ilyasfoo** [reply]
Hey, @cre8tivediva! Thanks so much for sharing your feedback! However, the screencast link you've provided has a passphrase to access.

> How can the message on the Shop page be modified? I know from my own audience of users, that will be the first thing they look for or send a message seeking how to change. I offer tech support and provide theme support and I can pretty much bet that will be one of the first questions a user will contact support for.

Good question! The can be edited when you're using a FSE theme, by going to `Apperance > Editor > Templates > Page: Coming soon`. Currently, the template to be edited will depend on your settings for `Restrict to store pages only`.

<img width="1374" alt="image" src="https://github.com/woocommerce/woocommerce/assets/3747241/45e7f43a-3402-4fea-a1b1-7bc2bb2779ad">

This information should be included in our documentation when it's ready.

> So if that was left there and they turned on Coming Soon mode, shouldn't these products disappear? Shouldn't there be a message like on the Shop page?

That's a great question. So far we haven't decided on having restrictions on blocks-level elements. My personal opinion is that it can be a little overreaching, and it may not be technically simple to do.

**Comment by cre8tivediva** [comment]
On the Private Link, it works, however, when I toggle it off, the Save button doesn't activate. When I go to he site incognito that link is still Live. Should the Save button light up to save turning off the Preview? I cleared the cache in the site and browser and the Shop still appears. I think I need to be able to save it.

![Screenshot 2024-05-22 at 19-29-56 WooCommerce settings ‹ Savannah Demo — WordPress](https://github.com/woocommerce/woocommerce/assets/4140035/337744f7-38e0-4d62-8ce6-4cafa4ebae83)

**Reply by adrianduffell** [reply]
Hmm, the save button should light up after toggling the private link or any of the settings. Could you please share more details about your environment (browser, operating system, host) and we'll investigate further.

**Reply by cre8tivediva** [reply]
So this is interesting. I've been testing in Firefox with Enhanced Tracking Protection "off". The button wouldn't light up. I just went to Chrome, logged in. Turned the link on, Saved. Then toggle it off. Saved. I went back over to Firefox, toggled and it wouldn't activate. I refreshed the page toggled it on and the button lit up, then toggled it off and the button lit up. I wonder if there's some other caching? This site is on PHP 7.4.33, so I might need to restore this site from before I started tinkering with it. Bump up to 8.2 to see if it still happens. I'll keep you posted.

**Reply by cre8tivediva** [reply]
I started over, bumped up to 8.2. I didn't have the button issue in Firefox anymore. However, in Firefox the cache won't flush so the Shop stays on even though Coming Soon is on. I even flushed my DNS through terminal. But it works right in Chrome.

Firefox 125.0.3 (64-bit) (needs an update)
Chrome 24.0.6367.208 (Official Build) (64-bit)

SiteGround hosting

**Comment by cre8tivediva** [comment]
@adrianduffell I just created a brand new staging on WPE. And I noticed after going back and forth toggling on and off settings, that there's caching going on. I am Live testing in Firefox but checking the site in FireFox Incognito. And when I toggle the settings - I have to clear the WPE caching for the pages to go offline or to come back online. More testing might be needed on various hosting platforms. Will there be a way to cache exclude any functionality in this? I'm going to turn off "object cache" to see if that helps.

**Reply by cre8tivediva** [reply]
I added these to the cache exclusions:

^/shop-2/
^/cart-2/

And they worked without clearing the cache, but unfortunately the home page was still displaying when I turned coming soon on the entire site. I still needed to clear the WP Engine cache for that to go offline.

**Reply by ilyasfoo** [reply]
Thanks so much for testing this out! In our current state, we have not considered cache busting mechanism on coming soon mode settings change. We'll look into this further.

Edit: Actually, we do have a [cache-busting mechanism](https://github.com/woocommerce/woocommerce/blob/5dd7713346786c5874615b34a1d56f81a29b673f/plugins/woocommerce/src/Internal/ComingSoon/ComingSoonCacheInvalidator.php#L24-L42) in place. We'll look into why it's not working, I suspect that hosts may have their own way to invalidate caches.

**Comment by thisissandip** [comment]
With themes not using the Site Editor, when we click on the "Editor" link to customize the landing page, it shows an error:

![fp2zKo.png](https://github.com/woocommerce/woocommerce/assets/69596988/c2e90758-3e01-464d-ab81-9a27034d634f)

And, in the frontend, we see the "Coming Soon" template is added along with the header and footer:

![ZNUP6o.png](https://github.com/woocommerce/woocommerce/assets/69596988/252874f7-99ba-4268-afa8-508fd90871d1)

In that case, the only option is to either use a theme with Site Editor to edit the template or not to use the Site Visibility Settings.

**Reply by psealock** [reply]
>With themes not using the Site Editor, when we click on the "Editor" link

Which editor link did you click? It shouldn't be visible from the side navigation:

<img width="377" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1922453/26885bd8-6465-4b96-a0fa-451346af52e9">

Nor would it be available from the homescreen badge:

<img width="314" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1922453/b1ef4b5d-4242-432b-9c69-7ff127e15e66">

**Reply by thisissandip** [reply]
I had the link open while using the TT4 theme and tried to refresh with SF theme active that's when I came across that.

<img width="712" alt="image" src="https://github.com/woocommerce/woocommerce/assets/69596988/eca47a93-cd5f-4ac6-a741-2ae804385c3b">

For stores using themes like SF, do users need to switch to a Site Editor-compatible theme to edit the "Coming Soon" template?

**Reply by psealock** [reply]
>For stores using themes like SF, do users need to switch to a Site Editor-compatible theme to edit the "Coming Soon" template?

yes, thats right.

**Reply by Brianmitchtay** [reply]
@psealock We'll document that sites using classic themes aren't able to edit the coming soon template. (Is there a template file they can override to make changes on a classic theme?)

@thisissandip 's comment got me looking closer at this and the experience right now on my test site. I'm seeing a bit of an odd edge case that _could_ be worth addressing. 

Recently I was working on the merchant documentation for all the WooCommerce settings, and I learned that if no shop page is assigned, but a page named `/shop/` exists, it will still get the shop template and be used as a product archive, despite not being assigned ~(now I'm actually wondering if maybe that's a bug technically?)~ Confirmed this is expected behavior

In any case, if that edge case is encountered, and the coming soon mode is restricted to store pages only, the coming soon template doesn't get applied to that page. It happens on block and classic themes.

I made a recording to try and explain more clearly. hopefully it's helpful.

https://github.com/woocommerce/woocommerce/assets/45057772/16df91f9-b47f-4c0b-afc4-1fbc070a7d3a

**Reply by ilyasfoo** [reply]
> We'll document that sites using classic themes aren't able to edit the coming soon template. (Is there a template file they can override to make changes on a classic theme?)

@Brianmitchtay There's no template file they can override to make changes on a classic theme. I don't think we can support it for now.

> In any case, if that edge case is encountered, and the coming soon mode is restricted to store pages only, the coming soon template doesn't get applied to that page. It happens on block and classic themes.

Closing the loop on this, it's fixed in https://github.com/woocommerce/woocommerce/pull/48522

**Comment by Brianmitchtay** [comment]
Catching another issue here. In my testing I'm seeing that when the "Restrict to store pages only" option is used, the privacy policy page is also restricted. AFAIK this is a Core WP setting and not actually a WooCommerce store page, despite being used by WooCommerce.

Should we be restricting this page from view? It seems that an established site that wants to restrict store pages may want to leave their privacy policy available for their users.

**Reply by Brianmitchtay** [reply]
For context as to why this stands out to me as a page we probably **shouldn't** be hiding as a store page, here's the section of documentation I was writing that sparked this comment.

---
With this option enabled, WooCommerce store pages that will display the "Coming Soon" page are:
- Cart page (Assigned at **WooCommerce > Settings > Advanced**)
- Checkout page (Assigned at **WooCommerce > Settings > Advanced**)
- Terms and conditions (Assigned at **WooCommerce > Settings > Advanced**)
- Shop page (Assigned at **WooCommerce > Settings > Products > General**)
- Privacy Policy Page (Assigned at **Settings > Privacy**)

**Reply by ilyasfoo** [reply]
Thanks for the feedback, @Brianmitchtay! Appreciate it.

@woocommerce/ghidorah I think we haven't addressed this. I agree that we should exclude privacy policy from store-only restriction. What do you think?

**Reply by ilyasfoo** [reply]
>  we should exclude privacy policy from store-only restriction. What do you think?

Circling back, this was addressed in https://github.com/woocommerce/woocommerce/pull/50608

**Comment by PoppyDev** [comment]
I have also noticed (which might be how I have my templates setup on elementor) that any pages with slug /products/ in that are WP core and not part of WooCommerce are also pushed to the 'coming Soon' page when activated.

Anyone have an idea how to get around this? Currently have to leave it as 'Live' to bring them back. Not a big issue but didn't realise after a good few days that it was effecting and has probably impacted on SEO and page range.

I would suggest allowing WooCommerce pages to be chosen from a dropdown or tick box, eliminating any core WP pages being targeted.

As seen by others, when you choose your Privacy Policy for WooCommerce, this also is targeted when its WP core. I understand if you had a separate Privacy Policy for your shop page but 90% of users probably merge this data into one property for ease of use and SEO.

Another issue is the 'Live' and 'Coming soon' tab next to your home (site url) tab in WP admin - top left. It look like your site as a whole is live / coming soon etc. I would only place this on the WooCommerce pages in WP Admin i.e. products, orders, woo settings etc.

Like the idea but its not ready to be part of a live build.

**Reply by ilyasfoo** [reply]
Thank you for the feedback, @PoppyDev!

> I have also noticed (which might be how I have my templates setup on elementor) that any pages with slug /products/ in that are WP core and not part of WooCommerce are also pushed to the 'coming Soon' page when activated.

This is unintended behaviour, I was able to reproduce it and filed an issue to address the bug: https://github.com/woocommerce/woocommerce/issues/51535

> Anyone have an idea how to get around this? Currently have to leave it as 'Live' to bring them back. Not a big issue but didn't realise after a good few days that it was effecting and has probably impacted on SEO and page range.

If you're comfortable with custom codes, you could use the `woocommerce_coming_soon_exclude` filter to exclude certain pages. In the following example, I'm excluding the page with ID `10` from coming soon. If you have multiple pages to exclude such as parent & child pages, you could try doing a check using base URL, such as `/products`.

```php
add_filter( 'woocommerce_coming_soon_exclude', function( $is_excluded ) {
    if ( get_the_ID() === 10 ) {
        return true;
    }
    return $is_excluded;
}, 10 );
```

> As seen by others, when you choose your Privacy Policy for WooCommerce, this also is targeted when its WP core. I understand if you had a separate Privacy Policy for your shop page but 90% of users probably merge this data into one property for ease of use and SEO.

I assume you meant `Terms and Conditions` ([screenshot](https://github.com/user-attachments/assets/d8af8bc5-6939-4574-bbca-d3ab4c6bdd6b))? Good point, I think we need to consider this use case. cc @woocommerce/ghidorah if you have an opinion.

> Another issue is the 'Live' and 'Coming soon' tab next to your home (site url) tab in WP admin - top left. It look like your site as a whole is live / coming soon etc. I would only place this on the WooCommerce pages in WP Admin i.e. products, orders, woo settings etc.

Thanks for your feedback, we're currently collating feedback on this recent change, and will decide on how to customize the appearance on individual stores in the upcoming releases.
